### ISSUE TITLE
DateTime with custom template are not work

### ISSUE THREAD

**Issue by mikasoukhov** [scottrippey/SmartFormat.NET#50]
Smart.Format("{0:HH:mm:ss}", DateTime.Now) == "" instead of valid string with current time

**Comment by scottrippey** [comment]
You must escape the colons.  This is the same behavior as String.Format.
```
Smart.Format("{0:HH\:mm\:ss}", DateTime.Now)
```

**Comment by scottrippey** [comment]
Unfortunately, in "Release" mode, the default behavior is to ignore errors.  That's why this works, but gives `""`.
This behavior is very confusing for many people, and hard to debug, especially since NuGet delivers "Release" packages.  I've been thinking about changing the default behavior, so I just created #51 to handle this.

**Comment by SeriousM** [comment]
Make it configurable, some users might already expect that behavior.

**Comment by scottrippey** [comment]
Well, it's already configurable. But I think the default config has been causing a lot of difficulty. It would definitely need a version bump.

<div>-------- Original message --------</div><div>From: Bernhard Millauer <notifications@github.com> </div><div>Date:03/18/2015  5:05 PM  (GMT-08:00) </div><div>To: "scottrippey/SmartFormat.NET" <SmartFormat.NET@noreply.github.com> </div><div>Cc: Scott Rippey <kokanut82@hotmail.com> </div><div>Subject: Re: [SmartFormat.NET] DateTime with custom template are not work (#50) </div><div>
</div>
Make it configurable, some users might already expect that behavior.


---

**Comment by mikasoukhov** [comment]
Console.WriteLine(string.Format("First =		{0:HH:mm:ss}", DateTime.Now));
Console.WriteLine(Smart.Format(@"Second =	{0:HH\:mm\:ss}", DateTime.Now));
Console.WriteLine(Smart.Format("Third =		{0:HH:mm:ss}", DateTime.Now));

In release mode output will be 

First = 10:41:51
Second =
Third =

I debug I've got an exception on second line (not suitable formatter).

**Comment by mikasoukhov** [comment]
Console.WriteLine(Smart.Format(@"Second =	{0:HH\:mm\:ss}", DateTime.Now));

**Comment by scottrippey** [comment]
Sorry, you're correct.  I was thinking of `TimeSpan`, which [does require escaping of `:`](https://msdn.microsoft.com/en-us/library/ee372287%28v=vs.110%29.aspx).

**Comment by scottrippey** [comment]
Possible solutions:
1. Change the syntax so it doesn't use double-colon.
  Eg. `{0,formatter:format}` or one of the other discussed options.
2. Make the named formatter feature smart -- if no formatter is found with that name, then assume it's not a named formatter.
  For example, look for a formatter `hh` with format `mm:ss`, and if it isn't found, use the default formatter with format `hh:mm:ss`.

**Comment by mhsimkin** [comment]
Hi, any resolution on this?  I'm just ran into this issue today when I tried to format at string like:

This job ran from {JobExecutionContext.PreviousFireTimeUtc:yyyy-MM-dd HH:mm:ss} - {JobExecutionContext.JobFireTimeUtc}.

I was given an unknown formatter error.

If I try to escape the colons then the code will not compile, as \: is an unrecognized escape code.

I've found a temporary work around, by specifying my form string as yyyy-MM-dd HH-mm-ss, however, I'm certain my users will not remember that and will also cause an error to be generated.

Thanks
marc

**Comment by mhsimkin** [comment]
Scott, I've spent some time reading through the various issues since the datetime formatting issue is a problem for me.
 
I like the idea of using a comma, just like .Net does for alignment.

Do you know when you are planning on implementing that change?

Another thought, maybe have some sort of configuration option that will enable/disable named formatters?

Thanks

marc

**Comment by scottrippey** [comment]
Marc,
Here's a few ideas...
First, what version of SmartFormat.NET are you using?  I've been testing these formatting issues using 1.6, and it seems to be working correctly: https://dotnetfiddle.net/bbAqbp  But when I switch to 1.5, I do get the "empty output" error.  
I don't specifically remember, but I think 1.6 removed the "unknown formatter" exception.  If the "formatter name" doesn't match anything, then it understands that it's not a formatter name.

Second, if you're stuck on 1.5 for some reason, you could try a `::` workaround.  Basically, you could just write `{Now::HH:mm:ss}` and that'll fix it.  It understands that `::` just means "ignore the named placeholder".

Third, I'm truly sorry for implementing the "named formatter" feature without properly vetting the breaking consequences.  I spent a lot of time struggling over the syntax, and I really thought the "double colon" syntax would be fully backwards compatible, and I was wrong.  
There have been a couple of issues related to syntax conflicts.  Looking back, I think maybe the comma syntax would've been a safer choice.  And perhaps I'll make another breaking change and go back to that.  
That being said, I do find great value in the "named formatter" feature, and I really want to encourage its use, so I'll be sure to put a good effort into giving it a great syntax.

**Comment by mhsimkin** [comment]
Hi Scott:

Thank you for your response.  I'm using 1.6.1.0 I pulled it from NuGet on June 1st.

I did a pull of the latest code from GitHub and put together this little test app, to see if I could figure out what originally how to write an extension.  I decided against it.  Then I ran into this formatting issue.  So I put together the below code to test, and see if I could figure out what is going wrong.

```
using System;
using SmartFormat;

namespace ConsoleApplication1
{
    public class JobContext
    {
        public DateTimeOffset? PreviousFireTimeUtc { get; set; }
        public DateTimeOffset? JobFireTimeUtc { get; set; }
    }

    class Program
    {
        static void Main(string[] args)
        {
            try
            {
                JobContext a = new JobContext();
                a.PreviousFireTimeUtc = DateTimeOffset.UtcNow.AddDays(-1);
                a.JobFireTimeUtc = DateTimeOffset.UtcNow;

                var msg = "This job ran from {JobExecutionContext.PreviousFireTimeUtc:yyyy-MM-dd HH:mm:ss} - {JobExecutionContext.JobFireTimeUtc}.";

                var data = new
                {
                    JobExecutionContext = a,
                    JobDetails = new {JobKey = new {Group = "Group1", Name = "Name1"}}
                };

                var x = Smart.Format(msg, data);
                Console.WriteLine(x);

            }
            catch (Exception ex)
            {
                Console.WriteLine(ex);
            }
        }
    }
}
```
I gave the double colon a try. I no longer get the format exception, which is good. However, there is now an additional colon in the output.  In this case right before the year.

```
This job ran from :2015-06-14 19:43:31 - 6/15/2015 7:43:31 PM +00:00.
```

Thanks

marc

**Comment by mhsimkin** [comment]
Scott, here is another thought.  If I'm not mistaken, from reading the list of issues, it is really the datetime and maybe timespan that have the most issues with the named formatter logic.  Maybe the solution is to just provide out of the box named formatters for datetime and timespan.  Then just indicate to people to use them.

As for myself, I need to figure out how to teach my users, how to do formatting of datetimes.  I'm thinking of just adding some sort of UI operation that they can choose a bunched of predetermined formats. I just need to think this through, but I need to know which way to call the Smart Format library.

Thanks for all the help.

marc

**Comment by scottrippey** [comment]
Regarding named formatters for DateTime and Timespan, you can actually use `{0:default:HH:mm:ss}` too.  `default` should handle them appropriately.

**Comment by scottrippey** [comment]
Do you think you could create a DotNetFiddle to reproduce the issue?  I couldn't recreate it.

**Comment by mhsimkin** [comment]
I can try, I don't know how to use DotNetFiddle.  I can also email you a VS 2013 solution, if that would help.  

Just let me know or send me a pvt email with your email.

Cheers

marc

**Comment by mhsimkin** [comment]
Scott, I tried to update the DotNetFiddle at the original link you provided.  Please let me know if you see the changes. BTW, everything works fine on the fiddle.

I did notice that the fiddle and my development machine are running different versions of the .NET 4.5 framework.

The DotNetFiddle is running version 4.0.30319.18449 of .Net 4.5.

My development and production systems are running: 4.0.30319.34209 . I wonder if something else changed in the patch.

marc

**Comment by scottrippey** [comment]
Should be fixed with #65

**Comment by axunonb** [comment]
Will be closed in upcoming version 1.7.x shortly.

**Comment by axunonb** [comment]
Updated version committed, where test with the issues mentioned have passed: https://github.com/scottrippey/SmartFormat.NET/commit/46bbae26b6c34582887b60fc0488cd134514fde7
